src/scvi/dataloaders/_anncollection.py

```diff
@@ -0,0 +1,174 @@
+import logging
+import math
+
+import anndata
+import numpy as np
+from scipy import sparse
+
+logger = logging.getLogger(__name__)
+
+
+class ArrayFakerSparse(sparse.csr_matrix):
+    """Create a wrapper around a sparse matrix layer in an AnnCollection.
+
+    This wrapper ensures compatibility with the typical AnnData API when
+    accessing or manipulating sparse matrix layers.
+    """
+
+    def __init__(self, collection, key):
+        self.collection = collection
+        self.key = key
+        # minibatch size for batchwise operations
+        self.batch_size = 256
+
+    def __getitem__(self, idx):
+        """Second layer of `__getitem__` that handles returning an index into a specific layer."""
+        return self.collection[idx].layers[self.key]
+
+    def getnnz(self):
+        """Run through the data in minibatches and count non-zero elements."""
+        n_batches = int(math.ceil(self.collection.shape[0] / self.n_batches))  # what is n_batches
+
+        nnz = 0
+
+        start = 0
+        end = self.batch_size
+        for _i in range(n_batches):
+            batch = self.collection[start:end].layers[self.key]
+            nnz += batch.getnnz()
+            start = end
+            end = min(end + self.batch_size, self.collection.shape[0])
+
+        return nnz
+
+    def _mul_vector(self, other):
+        """Included to try and trick issparse"""
+        raise NotImplementedError("This should basically never be called.")
+        return self._data.dot(other)
+
+    def _mul_multivector(self, other):
+        """Included to try and trick issparse"""
+        raise NotImplementedError("This should basically never be called.")
+        return self._data.dot(other)
+
+    def toarray(
+        self,
+    ):
+        """Convert to a dense array.
+
+        Notes
+        -----
+        Ths would require full dense representation in memory. It shouldn't be
+        used.
+        """
+        raise NotImplementedError("This should basically never be called.")
```

Comment:
Improve error message. Would this fail? I could see calling it on a subset of the data and it shouldn't fail then?